sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java

```diff
@@ -164,12 +166,40 @@ public boolean matches(String fileName) {
 
       public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channel)
         throws IOException {
-        ZipInputStream zip = new ZipInputStream(Channels.newInputStream(channel));
+        FullZipInputStream zip = new FullZipInputStream(Channels.newInputStream(channel));
         return Channels.newChannel(zip);
       }
     };
 
     /**
+     * Extend of {@link ZipInputStream} to automatically read all entries in the zip.
+     */
+    public class FullZipInputStream extends ZipInputStream {
+
+      private ZipEntry currentEntry;
+
+      public FullZipInputStream(InputStream is) throws IOException {
+        super(is);
+        currentEntry = getNextEntry();
+      }
+
+      public int read(byte[] b, int off, int len) throws IOException {
```

Comment:
This is the only read that has to be implemented, all the others are just wrappers for this read method.

int read() is the only one that you may want to handle specially since its very inefficient if its not implemented but its really inefficient for people to use so having an efficient implementation is useful.